validator/sawtooth_validator/concurrent/threadpool.py

```diff
@@ -0,0 +1,114 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+import logging
+import multiprocessing
+import time
+import os
+
+from concurrent.futures import ThreadPoolExecutor
+
+from sawtooth_validator.concurrent import atomic
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class InstrumentedThreadPoolExecutor(ThreadPoolExecutor):
+
+    def __init__(self, max_workers=None, name='', trace=None):
+        if trace is None:
+            if 'SAWTOOTH_TRACE_LOGGING' in os.environ:
+                self._trace = True
+            else:
+                self._trace = False
+        else:
+            self._trace = trace
+
+        self._name = name
+        if name == '':
+            self._name = 'Instrumented'
+
+        LOGGER.debug(
+            'Creating thread pool executor %s', self._name)
+
+        self._workers_in_use = atomic.Counter()
+
+        self._max_workers = max_workers
+        if self._max_workers is None:
+            # This is the same default as ThreadPoolExecutor, but we want to
+            # know how many workers there are for logging
+            self._max_workers = multiprocessing.cpu_count() * 5
+        super().__init__(max_workers)
+
+    def submit(self, task, *args, **kwargs):
+        submitted_time = time.time()
+
+        try:
+            task_name = task.__qualname__
+        except AttributeError:
+            task_name = str(task)
+
+        if self._trace:
+            task_details = \
+                str(task) + '[' + str(args) + ',' + str(kwargs) + ']'
+
+        def wrapper():
+            try:
+                start_time = time.time()
+                workers_already_in_use = self._workers_in_use.get_and_inc()
+                time_in_queue = (start_time - submitted_time) * 1000.0
+
+                if self._trace:
+                    LOGGER.debug('(%s) Task \'%s\' in queue for %0.3f ms',
+                        self._name,
+                        task_name,
+                        time_in_queue)
+                    LOGGER.debug(
+                        '(%s) Workers already in use %s/%s',
+                         self._name,
+                         workers_already_in_use,
+                         self._max_workers)
+                    LOGGER.debug(
+                        '(%s) Executing task %s', self._name, task_details)
+
+                return_value = task(*args, **kwargs)
+
+                end_time = time.time()
+                run_time = (end_time - start_time) * 1000.0
+
+                worker = self._workers_in_use.get_and_dec()
```

Comment:
Should probably move this to outside the try/catch so that it always gets decremented.